drivers/watchdog/wdt_sam0.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2017 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <watchdog.h>
+
+/* Device constant configuration parameters */
+struct wdt_sam0_dev_cfg {
+	Wdt *regs;
+};
+
+#define DEV_CFG(dev)							       \
+	((const struct wdt_sam0_dev_cfg *const)(dev)->config->config_info)
+
+static void wait_synchronization(Wdt *const wdt)
+{
+	while (wdt->STATUS.bit.SYNCBUSY) {
+	}
+}
+
+static void wdt_sam0_enable(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	wdt->CTRL.reg = WDT_CTRL_ENABLE;
+	wait_synchronization(wdt);
+}
+
+static void wdt_sam0_disable(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	wdt->CTRL.reg = 0;
+	wait_synchronization(wdt);
+}
+
+static int wdt_sam0_set_config(struct device *dev, struct wdt_config *config)
+{
+	return -ENOTSUP;
+}
+
+static void wdt_sam0_get_config(struct device *dev, struct wdt_config *config)
+{
+}
+
+static void wdt_sam0_reload(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	wdt->CLEAR.bit.CLEAR = WDT_CLEAR_CLEAR_KEY_Val;
+}
+
+static const struct wdt_driver_api wdt_sam0_api = {
+	.enable = wdt_sam0_enable,
+	.disable = wdt_sam0_disable,
+	.get_config = wdt_sam0_get_config,
+	.set_config = wdt_sam0_set_config,
+	.reload = wdt_sam0_reload,
+};
+
+static int wdt_sam0_init(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	/* Enable APB clock */
+	PM->APBAMASK.bit.WDT_ = 1;
```

Comment:
This will likely not work when we add another member of SAM0 family (D10, C20, etc.). At least for SAM family peripheral numbers do not match between different series / peripherals may be controlled by a different register. We probably need some function which will take peripheral ID as an argument and guarantee to program correct PM register.

Reply:
I checked ASF for all other devices in the series (samb11  samc20  samc21  samd09  samd10  samd11  samd20  samd21  samda1  samha1  saml21  saml22  samr21  samr30) and they all use the same flag here, yay!